File: db_func.rb
module DQCCdb

  # config
  require 'config'
  include DQCCconfig

  # for database connectivity
  require 'mongoid'


  Mongoid.configure do |config|
    name = DQCCconfig.db_dqor_name
    host = DQCCconfig.db_dqor_host
    config.master = Mongo::Connection.new.db(name)
  end


  # for hash computation
  require 'digest/md5'


  class Job
    include Mongoid::Document
    store_in "drqueue_jobs"

    field :name, :type => String
    field :startframe, :type => Integer
    field :endframe, :type => Integer
    field :blocksize, :type => Integer
    field :renderer, :type => String
    field :scenefile, :type => String
    field :retries, :type => Integer
    field :owner, :type => String
    field :created_with, :type => String
    field :rendertype, :type => String
    field :send_email, :type => Boolean
    field :email_recipients, :type => String
    field :file_provider, :type => String
  end

  class User
    include Mongoid::Document
    store_in "drqueue_users"

    field :name, :type => String
    field :admin, :type => Boolean, :default => false
    field :beta_user, :type => Boolean, :default => true
  end

  class Rendersession
    include Mongoid::Document
    store_in "cloudcontrol_rendersessions"

    field :user, :type => String
    field :num_slaves, :type => Integer
    field :run_time, :type => Integer
    field :vm_type, :type => String, :default => 't1.micro'
    field :costs, :type => Float
    field :paypal_token, :type => String
    field :paypal_payer_id, :type => String
    field :paid_at, :type => DateTime
    field :time_passed, :type => Integer, :default => 0
    field :start_timestamp, :type => Integer, :default => 0
    field :stop_timestamp, :type => Integer, :default => 0
    field :overall_time_passed, :type => Integer, :default => 0
  end



  # return list of all jobs known to DQOR
  def fetch_job_list
    puts "DEBUG: fetch_job_list()"

    #db_connection

    return Job.find(:all)
  end


  # return list of all rendersessions
  def fetch_rendersession_list
    puts "DEBUG: fetch_rendersession_list()"

    # fetch all paid and owned rendersessions
    sessions = Rendersession.all(:conditions => { :paid_at.ne => nil, :paypal_payer_id.ne => nil })
    puts "DEBUG: Rendersessions found: "+sessions.length.to_s
    return sessions
  end


  # return info about job owner
  def fetch_user_data(job_id)
    puts "DEBUG: fetch_user_data("+job_id.to_s+")"

    job = Job.find(job_id)
    return User.find(job.owner)
  end


  # return list of jobs belonging to a rendersession
  def fetch_rendersession_job_list(rendersession_id)
    puts "DEBUG: fetch_rendersession_job_list("+rendersession_id.to_s+")"

    rs = Rendersession.find(rendersession_id)
    return Job.all(:conditions => {:owner => rs.user})
  end


  # return active rendersessions of user
  def find_rendersession(user_hash)
    puts "DEBUG: find_rendersession("+user_hash+")"

    #db_connection

    needed_pm = nil
    payments = Payment.find(:all)
    payments.each do |pm|
      profile = Profile.find(pm.profile_id)
      profile_hash = Digest::MD5.hexdigest(profile.ldap_account)
      if profile_hash == user_hash
        needed_pm = pm.id
        break
      end
    end

    active_rs = nil
    Rendersession.find_all_by_payment_id(needed_pm).each do |rs|
      # check if there is time left
      if rs.time_passed < (rs.run_time * 3600 + rs.start_timestamp)
        # return only one session
        active_rs = rs
        break
      end
    end

    return active_rs
  end


end

File: db_func.py
# text coloring
from termcolor import colored

# config shared accross all modules/classes
import config as DQCCconfig

# modules imports shared accross all modules/classes
import global_imports as DQCCimport

# for database connectivity
from mongoengine import *

# Job objects stored in MongoDB
class Job(Document):
    meta = {
        "collection": "drqueue_jobs",
        "allow_inheritance": False,
    }
    name = StringField()
    startframe = IntField()
    endframe = IntField()
    blocksize = IntField()
    renderer = StringField()
    scenefile = StringField()
    retries = IntField()
    owner = StringField()
    created_with = StringField()
    rendertype = StringField()
    send_email = BooleanField()
    email_recipients = StringField()
    file_provider = StringField()


# User objects stored in MongoDB
class User(Document):
    meta = {
        "collection": "drqueue_users",
        "allow_inheritance": False,
    }
    name = StringField()
    admin = BooleanField(default=False)
    beta_user = BooleanField(default=True)


# Rendersession objects stored in MongoDB
class Rendersession(Document):
    meta = {
        "collection": "cloudcontrol_rendersessions",
        "allow_inheritance": False,
    }
    user = StringField()
    num_slaves = IntField()
    run_time = IntField()
    vm_type = StringField(default="t1.micro")
    costs = FloatField()
    active = BooleanField()
    paypal_token = StringField()
    paypal_payer_id = StringField()
    paid_at = DateTimeField()
    time_passed = IntField(default=0)
    start_timestamp = IntField(default=0)
    stop_timestamp = IntField(default=0)
    overall_time_passed = IntField(default=0)


# Computer objects stored in MongoDB
class Computer(Document):
    meta = {
        "collection": "drqueue_computers",
        "allow_inheritance": False,
    }
    engine_id = IntField()
    created_at = IntField()
    hostname = StringField()
    arch = StringField()
    os = StringField()
    proctype = StringField()
    nbits  = IntField()
    procspeed  = StringField()
    ncpus = IntField()
    ncorescpu = IntField()
    memory = FloatField()
    load = StringField()
    address = StringField()
    pools = ListField()


class DQCCdb():

    # debug config
    print(colored("\nDB configuration:", 'yellow', attrs=['reverse']))
    print("db_dqor_host = " + str(DQCCconfig.db_dqor_host))
    print("db_dqor_name = " + str(DQCCconfig.db_dqor_name))

    # connect to DB
    try:
        connect(DQCCconfig.db_dqor_name, host=DQCCconfig.db_dqor_host)
    except ConnectionError:
        print("\nCould not connect to MongoDB on host '" + DQCCconfig.db_dqor_host + "'! Check value of 'db_dqor_host' in dqcc.cfg.")
        exit(1)


    # return list of all rendersessions
    @staticmethod
    def fetch_rendersession_list():
        print(colored("DEBUG: DQCCdb.fetch_rendersession_list()", 'green'))
        # fetch all paid, owned and active rendersessions
        sessions = Rendersession.objects(paid_at__ne=None, paypal_payer_id__ne=None, active=True)
        print(colored("INFO: Rendersessions found: " + str(len(sessions)), 'yellow'))
        return sessions


    # return list of all active rendersessions (associated with running jobs)
    @staticmethod
    def fetch_active_rendersession_list():
        print(colored("DEBUG: DQCCdb.fetch_active_rendersession_list()", 'green'))
        # fetch all paid, owned and active rendersessions
        sessions = Rendersession.objects(paid_at__ne=None, paypal_payer_id__ne=None, active=True)
        active_sessions = []
        for rs in sessions:
            # fetch list of all belonging jobs
            all_jobs = DQCCdb.fetch_rendersession_job_list(rs)
            active_jobs = DQCCdb.fetch_rendersession_active_job_list(rs)
            # skip this rendersession if not used
            if len(all_jobs) == 0:
                print(colored("INFO: Rendersession " + str(rs.id) + " isn't in use yet (has no associated jobs). Skipping this one.", 'yellow'))
                # skip to next session
                continue
            elif len(active_jobs) > 0:
                active_sessions.append(rs)
            else:
                # remove eventually running slaves of this session
                if (len(active_jobs) == 0) and (len(all_jobs) > 0):
                    # update time counter
                    if rs.stop_timestamp == 0:
                        rs.overall_time_passed += rs.time_passed
                        rs.time_passed = 0
                        rs.stop_timestamp = int(time.time())
                        print(colored("INFO: Setting stop timestamp to: " + str(rs.stop_timestamp) + ".", 'yellow'))
                        rs.save()
                else:
                    # skip to next session
                    continue
        print(colored("INFO: Active rendersessions found: " + str(len(active_sessions)), 'yellow'))
        return active_sessions


    # return list of jobs belonging to a rendersession
    @staticmethod
    def fetch_rendersession_job_list(rendersession):
        print(colored("DEBUG: DQCCdb.fetch_rendersession_job_list(" + str(rendersession) + ")", 'green'))
        # search for all jobs of the rendersession's owner
        jobs = Job.objects(owner=rendersession.user)
        return jobs


    # return list of active jobs belonging to a rendersession
    @staticmethod
    def fetch_rendersession_active_job_list(rendersession):
        print(colored("DEBUG: DQCCdb.fetch_rendersession_active_job_list(" + str(rendersession) + ")", 'green'))
        # search for all jobs of the rendersession's owner
        jobs = Job.objects(owner=rendersession.user)
        running_jobs = []
        # fetch info for each job and check status
        for job in jobs:
            job_info = DQCCimport.DQCCqueue.fetch_job_info(job.id)
            if job_info == None:
                print(colored("ERROR: Queue info for job " + str(job.id) + " could not be fetched.", 'red'))
            else:
                # see if there is any job active or waiting
                if DQCCimport.DQCCqueue.job_status(job.id) == "pending":
                    running_jobs.append(job)
        return running_jobs


    # search for computer by address
    @staticmethod
    def query_computer_by_address(comp_address):
        print(colored("DEBUG: DQCCdb.query_computer_by_address(" + str(comp_address) + ")", 'green'))
        try:
            # get data of first element as dict
            computer = Computer.objects(address=comp_address)[0]._data
            print(colored("INFO: Computer found: engine_id " + str(computer["engine_id"]), 'yellow'))
        # first element of list is not accessible when nothing was found
        except IndexError:
            print(colored("INFO: Computer not found", 'yellow'))
            computer = None
        return computer
